source/gui/settingsDialogs.py

```diff
@@ -928,144 +928,193 @@ def onOk(self, evt):
-		slider.SetValue(getattr(getSynth(),setting.name))
-		sizer.Add(label)
-		sizer.Add(slider)
+		labeledControl = guiHelper.LabeledControlHelper(
+			self,
+			"%s:"%setting.displayNameWithAccelerator,
+			nvdaControls.EnhancedInputSlider,
+			minValue=setting.minVal,
+			maxValue=setting.maxVal
+		)
+		lSlider=labeledControl.control
+		setattr(self,"%sSlider"%setting.name,lSlider)
+		lSlider.Bind(wx.EVT_SLIDER,DriverSettingChanger(self.driver,setting))
+		self._setSliderStepSizes(lSlider,setting)
+		lSlider.SetValue(getattr(self.driver,setting.name))
 		if self.lastControl:
-			slider.MoveAfterInTabOrder(self.lastControl)
-		self.lastControl=slider
-		return sizer
+			lSlider.MoveAfterInTabOrder(self.lastControl)
+		self.lastControl=lSlider
+		return labeledControl.sizer
 
 	def makeStringSettingControl(self,setting):
-		"""Same as L{makeSettingControl} but for string settings. Returns sizer with label and combobox."""
+		"""Same as L{makeSliderSettingControl} but for string settings. Returns sizer with label and combobox."""
 
 		labelText="%s:"%setting.displayNameWithAccelerator
-		synth=getSynth()
-		setattr(self,"_%ss"%setting.name,getattr(synth,"available%ss"%setting.name.capitalize()).values())
-		l=getattr(self,"_%ss"%setting.name)###
+		setattr(self,"_%ss"%setting.name,getattr(self.driver,"available%ss"%setting.name.capitalize()).values())
+		l=getattr(self,"_%ss"%setting.name)
 		labeledControl=guiHelper.LabeledControlHelper(self, labelText, wx.Choice, choices=[x.name for x in l])
 		lCombo = labeledControl.control
 		setattr(self,"%sList"%setting.name,lCombo)
 		try:
-			cur=getattr(synth,setting.name)
+			cur=getattr(self.driver,setting.name)
 			i=[x.ID for x in l].index(cur)
 			lCombo.SetSelection(i)
 		except ValueError:
 			pass
-		lCombo.Bind(wx.EVT_CHOICE,StringSynthSettingChanger(setting,self))
+		lCombo.Bind(wx.EVT_CHOICE,StringDriverSettingChanger(self.driver,setting,self))
 		if self.lastControl:
 			lCombo.MoveAfterInTabOrder(self.lastControl)
 		self.lastControl=lCombo
 		return labeledControl.sizer
 
 	def makeBooleanSettingControl(self,setting):
-		"""Same as L{makeSettingControl} but for boolean settings. Returns checkbox."""
+		"""Same as L{makeSliderSettingControl} but for boolean settings. Returns checkbox."""
 		checkbox=wx.CheckBox(self,wx.ID_ANY,label=setting.displayNameWithAccelerator)
 		setattr(self,"%sCheckbox"%setting.name,checkbox)
 		checkbox.Bind(wx.EVT_CHECKBOX,
-			lambda evt: setattr(getSynth(),setting.name,evt.IsChecked()))
-		checkbox.SetValue(getattr(getSynth(),setting.name))
+			lambda evt: setattr(self.driver,setting.name,evt.IsChecked()))
+		checkbox.SetValue(getattr(self.driver,setting.name))
 		if self.lastControl:
 			checkbox.MoveAfterInTabOrder(self.lastControl)
 		self.lastControl=checkbox
 		return checkbox
 
+	def updateDriverSettings(self, changedSetting=None):
+		"""Creates, hides or updates existing GUI controls for all of supported settings."""

```

Comment:
I expect there is a way we could avoid all this `isInstance` stuff. Probably not now, but it would be nice to see if there is some way to simplify this code.

Reply:
I guess we could move the setting makers to the driverHandler, i.e. make them methods on the driver settings. This also adds the ability for driver authors to create their own setting makers for specific settings, which would actually be very cool in scope of the gui of the vision framework. It would also solve the issue with the exception that is now applied to the voice setting.

Reply:
I think this should be pulled out, right now it's hard to read. What about the driver provides a factory for constructing these. The common ones can be pulled out of here and provided for re-use. 

For now, leave this as it is, in the interests of getting this merged.

Reply:
I agree